File: Labs/experiment2/G_grocery_store.py
n = int(input())
items = []
for _ in range(n):
    w, v, c = map(int, input().split())
    items.append((w, v, c))

items.sort(key=lambda x: x[2] * x[0])  # sort by c_i * w_i ascending

suf = [0] * (n + 2)
for i in range(n, 0, -1):
    suf[i] = suf[i + 1] + items[i - 1][0]

ans = 0
for i in range(1, n + 1):
    w, v, c = items[i - 1]
    ans += v - c * suf[i + 1]
print(ans)

File: Labs/experiment2/G_grocery_store.cpp
#include <iostream>
#include <algorithm>
using namespace std;

const int N = 2000001;
struct Node {
    long long w, v, c;
} a[N];

bool cmp(Node n1, Node n2) {
    return n2.c * n1.w < n1.c * n2.w;
}

long long suf[N];

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n;
    cin >> n;
    for (int i = 1; i <= n; i++) cin >> a[i].w >> a[i].v >> a[i].c;
    sort(a + 1, a + n + 1, cmp);
    for (int i = n; i >= 1; i--) suf[i] = suf[i + 1] + a[i].w;
    long long ans = 0;
    for (int i = 1; i <= n; i++) ans += a[i].v - a[i].c * suf[i + 1];
    cout << ans << endl;
    return 0;
}
